src/dataMap/metaManager/lazyFactoryMap.js

```diff
@@ -0,0 +1,315 @@
+import { arrayFilter } from '../../helpers/array';
+import { assert, isUnsignedNumber, isNullish } from './utils';
+
+/**
+ * The LazyFactoryMap object holds key-value pairs in the structure similar to the
+ * regular Map. Once created, items can be moved around a grid depending on the operations
+ * performed on that grid - adding or removing rows. The collection requires "key"
+ * to be a zero-based index.
+ *
+ * It's essential to notice that the "key" index under which the item was created
+ * is volatile. After altering the grid, the "key" index can change.
+ *
+ * Having created N items with corresponding example data where the data has 10
+ * holes (`undefined` values) within (that's why internal storage index counts from 10).
+ * +------+------+------+------+------+
+ * | 0/10 | 1/11 | 2/12 | 3/13 | 4/14 |  Keys (volatile zero-based index / internal storage index)
+ * +------+------+------+------+------+
+ *    │      │      │      │      │
+ * +------+------+------+------+------+
+ * | AAA  | BBB  | CCC  | DDD  | EEE  |  Data
+ * +------+------+------+------+------+
+ *
+ * map.obtain(0) // returns "AAA"
+ * map.obtain(2) // returns "CCC"
+ *
+ * after inserting new 2 rows, keys that hold the data positioned after the place
+ * where the new rows are added are upshifted by 2.
```

Comment:
Ohh, I am a bit lost in this sentence. `Upshifting` relates to the `keys`, right?

Reply:
Right! It's related to `key`. The key is floating, after inserting 2 new rows the data will be available under the key + 2.